app/src/main/java/bg/zahov/app/ui/workout/add/AddTemplateWorkoutViewModel.kt

```diff
@@ -1,368 +1,568 @@
-import android.view.View
-import androidx.lifecycle.AndroidViewModel
-import androidx.lifecycle.LiveData
-import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
-import bg.zahov.app.data.exception.CriticalDataNullException
+import bg.zahov.app.Inject
+import bg.zahov.app.data.interfaces.WorkoutProvider
+import bg.zahov.app.data.model.Category
 import bg.zahov.app.data.model.Exercise
 import bg.zahov.app.data.model.SetType
 import bg.zahov.app.data.model.Sets
-import bg.zahov.app.data.model.Units
-import bg.zahov.app.data.model.Units.METRIC
+import bg.zahov.app.data.model.ToastManager
 import bg.zahov.app.data.model.Workout
-import bg.zahov.app.getReplaceableExerciseProvider
-import bg.zahov.app.getSelectableExerciseProvider
-import bg.zahov.app.getServiceErrorProvider
-import bg.zahov.app.getSettingsProvider
-import bg.zahov.app.getWorkoutProvider
+import bg.zahov.app.data.provider.AddExerciseToWorkoutProvider
+import bg.zahov.app.data.provider.ReplaceableExerciseProvider
+import bg.zahov.app.ui.workout.WorkoutEntry
+import bg.zahov.app.ui.workout.filterDoubleInput
+import bg.zahov.app.ui.workout.toSetEntry
+import bg.zahov.app.ui.workout.toWorkoutEntry
+import bg.zahov.app.ui.workout.toWorkoutEntryList
 import bg.zahov.app.util.filterIntegerInput
 import bg.zahov.app.util.generateRandomId
-import bg.zahov.app.util.toExercise
-import bg.zahov.app.util.toExerciseSetAdapterSetWrapper
-import bg.zahov.app.util.toExerciseSetAdapterWrapper
 import bg.zahov.fitness.app.R
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.launch
-import java.lang.Exception
 import java.time.LocalDateTime
 
-class AddTemplateWorkoutViewModel(application: Application) : AndroidViewModel(application) {
-    private val workoutProvider by lazy {
-        application.getWorkoutProvider()
-    }
-
-    private val selectableExerciseProvider by lazy {
-        application.getSelectableExerciseProvider()
-    }
-
-    private val replaceableExerciseProvider by lazy {
-        application.getReplaceableExerciseProvider()
-    }
+/**
+ * Represents the UI state for adding a template workout in the application.
+ *
+ * @property workoutName The name of the workout being added. Defaults to an empty string.
+ * @property workoutDate The date and time of the workout. Must be provided when creating the instance.
+ * @property exercises A list of exercises (or workout entries) included in the template workout. Defaults to an empty list.
+ * @property note An optional note associated with the workout. Defaults to an empty string.
+ * @property isAdded A flag indicating whether the workout has been successfully added. Defaults to `false`.
+ */
+data class AddTemplateWorkoutUiState(
+    val workoutName: String = "",
+    val workoutDate: LocalDateTime,
```

Comment:
what is the purpose of this date ui state field. do you show it to the user, if not you can remove it from the state and set it when you need it to save your Workout. There is chance I missed the usage in the ui

Reply:
It is for whenever we save the workout to know when it was created. Removed it from the data class and simply added it as LocalDateTime.now() whenever we need it because it doesn't make much of a difference